File: code_java/src/main/java/jp/co/pm/ai/desktop/reconciliation/EcSideClassification.java
package jp.co.pm.ai.desktop.reconciliation;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * 依頼書の加工内容・EC面から両面EC/片面ECを判定する。
 *
 * <p>Python {@code planning_core.core.ec_side_classification} と同一ロジック。
 */
public final class EcSideClassification {

    public static final String COLUMN_TITLE = "EC面区分";
    public static final String DOUBLE_SIDED = "両面EC";
    public static final String SINGLE_SIDED = "片面EC";
    public static final String UNKNOWN = "不明";

    private static final List<String> BLANK_EC_MARKERS =
            List.of("-", "―", "－", "—", "nan", "none", "null");

    private EcSideClassification() {}

    public static String classify(String processContent, String ecMen) {
        return classify(processContent, ecMen, true, false);
    }

    public static String classify(String processContent, String ecMen, boolean juchuRowFound) {
        return classify(processContent, ecMen, juchuRowFound, false);
    }

    /**
     * 段階1向け判定。EC面 空かつ依頼書原本参照無しは {@link #UNKNOWN}。
     * EC面 空かつ原本参照ありは {@link #DOUBLE_SIDED}。
     */
    public static String classify(
            String processContent, String ecMen, boolean juchuRowFound, boolean originalRefFound) {
        if (!processContentHasEc(processContent)) {
            return "";
        }
        if (!juchuRowFound) {
            return UNKNOWN;
        }
        if (isBlankEcMen(ecMen)) {
            return originalRefFound ? DOUBLE_SIDED : UNKNOWN;
        }
        if (ecMenIndicatesDoubleSided(ecMen)) {
            return DOUBLE_SIDED;
        }
        if (ecMenIndicatesSingleSided(ecMen)) {
            return SINGLE_SIDED;
        }
        return UNKNOWN;
    }

    static boolean ecMenIndicatesDoubleSided(String ecMen) {
        if (isBlankEcMen(ecMen)) {
            return false;
        }
        String norm = JuchuTransferValueNormalizer.normalizeText(ecMen);
        return norm.contains("両面");
    }

    static boolean ecMenIndicatesSingleSided(String ecMen) {
        if (isBlankEcMen(ecMen) || ecMenIndicatesDoubleSided(ecMen)) {
            return false;
        }
        String norm = JuchuTransferValueNormalizer.normalizeText(ecMen);
        String upper = norm.toUpperCase(Locale.ROOT);
        if (norm.contains("片面")) {
            return true;
        }
        if ("H".equals(upper) || "Q".equals(upper)) {
            return true;
        }
        if (upper.startsWith("H面") || upper.startsWith("Ｈ面")) {
            return true;
        }
        if (upper.startsWith("Q面") || upper.startsWith("Ｑ面")) {
            return true;
        }
        if (norm.contains("ｽﾗｲｽ") || norm.contains("スライス")
                || norm.contains("ｽｷﾝ") || norm.contains("スキン")) {
            return true;
        }
        return !norm.isEmpty();
    }

    /**
     * 配台: 両面EC は EC 工程では2回分。SEC 工程はワンパスで両面EC 可能のため1回分。
     * 国分工場（{@code PM_AI_FACTORY_SITE=KOKUBU}）は1パスで両面EC 可能なため等倍。
     */
    public static int ecDispatchPassCount(String ecSideClass, String processName) {
        return ecDispatchPassCount(ecSideClass, processName, System.getenv("PM_AI_FACTORY_SITE"));
    }

    public static int ecDispatchPassCount(
            String ecSideClass, String processName, String factorySite) {
        if (!DOUBLE_SIDED.equals(ecSideClass != null ? ecSideClass.strip() : "")) {
            return 1;
        }
        if (isKokubuFactory(factorySite)) {
            return 1;
        }
        String proc = normalizeProcessName(processName);
        if ("SEC".equals(proc)) {
            return 1;
        }
        if ("EC".equals(proc)) {
            return 2;
        }
        return 1;
    }

    static boolean isKokubuFactory(String factorySite) {
        if (factorySite == null || factorySite.isBlank()) {
            return false;
        }
        return "KOKUBU".equalsIgnoreCase(factorySite.strip());
    }

    static String normalizeProcessName(String raw) {
        if (raw == null) {
            return "";
        }
        String t = JuchuTransferValueNormalizer.normalizeText(raw.strip());
        return t.replaceAll("[\\s　]+", "");
    }

    public static boolean processContentHasEc(String processContent) {
        for (String tok : parseProcessContentTokens(processContent)) {
            String upper = tok.toUpperCase(Locale.ROOT);
            if ("EC".equals(upper) || upper.startsWith("EC")) {
                return true;
            }
        }
        return false;
    }

    /**
     * 枝番依頼NO の EC 面 lookup 用親キー。
     * 例: {@code W7-22-1} → {@code W7-22}（末尾の {@code -} + 数字 1 セグメントを除く）。
     */
    public static String parentIraiNoLookupKey(String iraiNo) {
        String key = RequestFormOriginalIndexLookup.normalizeIraiNoKey(iraiNo);
        if (key.isEmpty() || !key.contains("-")) {
            return "";
        }
        int lastDash = key.lastIndexOf('-');
        if (lastDash <= 0 || lastDash >= key.length() - 1) {
            return "";
        }
        String tail = key.substring(lastDash + 1);
        if (tail.isEmpty() || !tail.chars().allMatch(Character::isDigit)) {
            return "";
        }
        String head = key.substring(0, lastDash);
        if (!head.contains("-")) {
            return "";
        }
        return head;
    }

    /** 依頼NO 直 lookup → 親依頼NO の順で EC面区分文字列を返す。 */
    public static String resolveEcSideClass(Map<String, String> byKey, String iraiNo) {
        if (byKey == null || byKey.isEmpty()) {
            return "";
        }
        String key = RequestFormOriginalIndexLookup.normalizeIraiNoKey(iraiNo);
        String direct = byKey.get(key);
        if (direct != null && !direct.isBlank()) {
            return direct.strip();
        }
        String parent = parentIraiNoLookupKey(iraiNo);
        if (!parent.isEmpty()) {
            String inherited = byKey.get(parent);
            if (inherited != null && !inherited.isBlank()) {
                return inherited.strip();
            }
        }
        return "";
    }

    static List<String> parseProcessContentTokens(String val) {
        if (JuchuTransferValueNormalizer.isBlank(val)) {
            return List.of();
        }
        String s = JuchuTransferValueNormalizer.normalizeText(val);
        if (s.isEmpty()) {
            return List.of();
        }
        String lower = s.toLowerCase(Locale.ROOT);
        if ("nan".equals(lower) || "none".equals(lower) || "null".equals(lower)) {
            return List.of();
        }
        List<String> out = new ArrayList<>();
        for (String part : s.split(",")) {
            String t = part != null ? part.strip() : "";
            if (!t.isEmpty()) {
                out.add(t);
            }
        }
        return List.copyOf(out);
    }

    private static boolean isBlankEcMen(String ecMen) {
        if (JuchuTransferValueNormalizer.isBlank(ecMen)) {
            return true;
        }
        String norm = JuchuTransferValueNormalizer.normalizeText(ecMen);
        return norm.isEmpty() || BLANK_EC_MARKERS.contains(norm.toLowerCase(Locale.ROOT));
    }
}

File: code/python/planning_core/core/ec_side_classification.py
# -*- coding: utf-8 -*-
"""依頼書の加工内容・EC面から両面EC/片面ECを判定する。"""
from __future__ import annotations

import os
import re
import unicodedata

EC_SIDE_CLASS_DOUBLE = "両面EC"
EC_SIDE_CLASS_SINGLE = "片面EC"
EC_SIDE_CLASS_UNKNOWN = "不明"

EC_DISPATCH_PASS_COUNT_DOUBLE = 2


def _is_blank(val) -> bool:
    if val is None:
        return True
    if isinstance(val, float):
        import math

        if math.isnan(val):
            return True
    s = unicodedata.normalize("NFKC", str(val).strip())
    return not s or s in {"-", "―", "－", "—", "nan", "none", "null"}


def parse_process_content_tokens(val) -> list[str]:
    if val is None:
        return []
    if isinstance(val, float):
        import math

        if math.isnan(val):
            return []
    s = unicodedata.normalize("NFKC", str(val).strip())
    if not s or s.lower() in ("nan", "none", "null"):
        return []
    return [p.strip() for p in s.split(",") if p.strip()]


def process_content_has_ec(val) -> bool:
    for tok in parse_process_content_tokens(val):
        upper = unicodedata.normalize("NFKC", tok).upper()
        if upper == "EC" or upper.startswith("EC"):
            return True
    return False


def ec_men_indicates_double_sided(ec_men) -> bool:
    """受注 EC面 列の「両面」等を両面EC と解釈する。"""
    if _is_blank(ec_men):
        return False
    s = unicodedata.normalize("NFKC", str(ec_men).strip())
    return "両面" in s


def ec_men_indicates_single_sided(ec_men) -> bool:
    """受注 EC面 列の H/Q/片面 等を片面EC と解釈する。"""
    if _is_blank(ec_men) or ec_men_indicates_double_sided(ec_men):
        return False
    s = unicodedata.normalize("NFKC", str(ec_men).strip())
    upper = s.upper()
    if "片面" in s:
        return True
    if upper in ("H", "Q"):
        return True
    if upper.startswith("H面") or upper.startswith("Ｈ面"):
        return True
    if upper.startswith("Q面") or upper.startswith("Ｑ面"):
        return True
    if "ｽﾗｲｽ" in s or "スライス" in s or "ｽｷﾝ" in s or "スキン" in s:
        return True
    return bool(s)


def classify_ec_side(
    process_content,
    ec_men,
    *,
    juchu_row_found: bool = True,
    original_ref_found: bool = False,
) -> str:
    """
    加工内容に EC 工程があり EC面 が「両面」なら両面EC。
    EC面 が H/Q/片面 等なら片面EC。
    EC面 が空で依頼書原本（親依頼NO含む）が参照できれば両面EC。
    受注 lookup 無し、または EC面 空かつ原本無しは「不明」。
    """
    if not process_content_has_ec(process_content):
        return ""
    if not juchu_row_found:
        return EC_SIDE_CLASS_UNKNOWN
    if ec_men_indicates_double_sided(ec_men):
        return EC_SIDE_CLASS_DOUBLE
    if _is_blank(ec_men):
        if original_ref_found:
            return EC_SIDE_CLASS_DOUBLE
        return EC_SIDE_CLASS_UNKNOWN
    if ec_men_indicates_single_sided(ec_men):
        return EC_SIDE_CLASS_SINGLE
    return EC_SIDE_CLASS_UNKNOWN


def _normalize_process_name_for_ec_dispatch(raw) -> str:
    t = unicodedata.normalize("NFKC", str(raw or "").strip())
    return re.sub(r"[\s　]+", "", t)


def _factory_site_is_kokubu(factory_site=None) -> bool:
    raw = factory_site
    if raw is None:
        raw = os.environ.get("PM_AI_FACTORY_SITE") or "KONAN"
    return str(raw or "").strip().upper() == "KOKUBU"


def ec_dispatch_pass_count(ec_side_class, machine=None, factory_site=None) -> int:
    """
    配台: 両面EC は EC 機械では EC 加工を2回分として扱う。
    SEC 機械はワンパスで両面EC 可能のため常に1回分。
    国分工場（PM_AI_FACTORY_SITE=KOKUBU）は1パスで両面EC 可能なため等倍（1回分）。
    """
    if str(ec_side_class or "").strip() != EC_SIDE_CLASS_DOUBLE:
        return 1
    if _factory_site_is_kokubu(factory_site):
        return 1
    proc = _normalize_process_name_for_ec_dispatch(machine)
    if proc == "SEC":
        return 1
    if proc == "EC":
        return EC_DISPATCH_PASS_COUNT_DOUBLE
    return 1
